File: Paint_House-BF.java
// 256. Paint House - https://leetcode.com/problems/paint-house/
// Time Complexity : O(2^N)
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :

class Solution{
    public int minCost(int[][] costs){
        if(costs == null || coins.length == 0)    
            return 0;

        //coins, row, column, minCost
        int case1 = helper(costs,0,0,0);
        int case2 = helper(costs,0,1,0);
        int case3 = helper(costs,0,2,0);

        return Math.min(case1. Math.min(case2,case3));
    }

    private int helper(int[][] costs, int row, int cloumn, int minCost)
    {
        // base
        if(row==costs.length) return minCost;

        if(color == 0)
        {
            return Math.min(
                helper(costs, row+1, 1, minCost + costs[row][0]), 
                helper(costs, row+1, 2, minCost + costs[row][0])
                );
        }
        else if(color == 1)
        {
            return Math.min(
                helper(costs, row+1, 0, minCost + costs[row][1]), 
                helper(costs, row+1, 2, minCost + costs[row][1])
                );
        }
        else
        {
            return Math.min(
                helper(costs, row+1, 1, minCost + costs[row][2]), 
                helper(costs, row+1, 0, minCost + costs[row][2])
                );
        }
    }
}

File: Paint_House_BF.py
# The solution below takes the BF approach
# However, it fails to perform on all the cases.
# Time complexity is Exponential 3^n and space complexity is exponential as well

class Solution:
    def minCost(self, costs: List[List[int]]) -> int:
        if not costs: return 0
        case1 = self.helper(costs,row=0, col=0,mincost=0)
        case2 = self.helper(costs,row=0, col=1,mincost=0)
        case3 = self.helper(costs,row=0, col=2,mincost=0)
        return min(case1,case2,case3)

    def helper(self,costs,row,col,mincost):
        # Base Case
        if row == len(costs): return mincost

        # Logic
        if col ==0:
            return min( self.helper(costs, row+1, col+1, mincost+costs[row][col]), self.helper(costs,row+1, col+2, mincost + costs[row][col]))
        if col ==1:
            return min( self.helper(costs, row+1, col-1, mincost+costs[row][col]), self.helper(costs,row+1, col+1, mincost + costs[row][col]))
        if col ==2:
            return min( self.helper(costs, row+1, col-1, mincost+costs[row][col]), self.helper(costs,row+1, col-2, mincost + costs[row][col]))
